### PR TITLE
Calling procedures from Cypher

### PR DESCRIPTION
**Pull request by pontusmelke**
Support for calling procedures from Cypher.
- Added built-in procedures
- Added new `ExecutablePlanBuilder` that takes on queries that doesn't require planning
- In the process added support for all schema operations in the new `ExecutablePlanBuilder`
- Clean-up removed external dependencies and unused code

### CODE REVIEW

**Comment by jakewins** [comment]
This is so exciting! I read through, could not find anything that raised any bells - but perhaps we could have a second pair of cypher-savvy eyes oogle over it before merging? /cc @systay

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/PureSideEffectExecutionPlan.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    val taskCloser = new TaskCloser
+    taskCloser.addTask(countingCtx.close)
+    if (planType == ExplainMode) {
+      //close all statements
+      taskCloser.close(success = true)
+      new ExplainExecutionResult(List.empty, description, queryType, Set.empty)
+    } else
+      sideEffect(countingCtx)
+      taskCloser.close(success = true)
+      PureSideEffectInternalExecutionResult(description, taskCloser, countingCtx, queryType, planType)
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/PureSideEffectExecutionPlan.scala]: Either indentation here is really off or the else block is wrong due to missing brackets.

Assuming that brackets are missing I guess that the call to `taskCloser.close` should be done by the result and not performed here.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/PureSideEffectExecutionPlan.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  extends ExecutionPlan {
+
+  override def run(ctx: QueryContext, planType: ExecutionMode,
+                   params: Map[String, Any]): InternalExecutionResult = {
+    val countingCtx = new UpdateCountingQueryContext(ctx)
+    val taskCloser = new TaskCloser
+    taskCloser.addTask(countingCtx.close)
+    if (planType == ExplainMode) {
+      //close all statements
+      taskCloser.close(success = true)
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/PureSideEffectExecutionPlan.scala]: Why do we need to create a `taskCloser` in this case? Looks unnecessary.

**Comment by Mats-SX** [comment]
Were all my comments disregarded??

EDIT: Stupid GitHub lost them because I commented on the commit rather than the PR. I'll repost them.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+import java.util.stream.Stream
+
+import org.neo4j.cypher.{CypherTypeException, ExecutionEngineFunSuite, InvalidArgumentException}
+import org.neo4j.kernel.api.KernelAPI
+import org.neo4j.proc.Procedure.{BasicProcedure, Context}
+import org.neo4j.proc.ProcedureSignature.procedureSignature
+import org.neo4j.proc.{Neo4jTypes, ProcedureSignature => KernelSignature}
+
+class CallProcedureAcceptanceTest extends ExecutionEngineFunSuite {
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:31]: Add a test for calling a non-existent procedure, and tests were argument values are `null`s.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+  test("should be able to find labels from built-in-procedure") {
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:33]: Add test for this call on an empty database, on a database with nodes but no labels, on a db which had a node with a label, but the label was removed, and when the whole node was removed.

Same for relationships and their types, and also properties.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+  test("should fail a procedure declares a float but gets an integer") {
+    // Given
+    register(Neo4jTypes.NTFloat)
+
+    // Then
+    a [CypherTypeException] shouldBe thrownBy(execute("CALL my.first.proc(42)"))
+  }
+
+  test("should fail if explicit argument is missing") {
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:250]: Add test for when too many arguments are sent (it requests two ints, and gets two ints and a string, or three ints, etc), and perhaps some of them with the same key/name.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/PlannerName.scala [hunk 6]
```diff
@@ -60,6 +60,13 @@ case object DPPlannerName extends CostBasedPlannerName {
   val name = "DP"
 }
 
+/**
+  * Procedures are not really planned, b
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/PlannerName.scala]: Unfinished comment.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/PlannerName.scala [hunk 7]
```diff
@@ -60,6 +60,13 @@ case object DPPlannerName extends CostBasedPlannerName {
...
+  */
+case object ProcedurePlannerName extends CostBasedPlannerName {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/PlannerName.scala]: This should not be a CostBasedPlannerName. Perhaps inherit directly from PlannerName?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+import org.neo4j.cypher.internal.frontend.v3_0.EntityNotFoundException
+import org.neo4j.cypher.internal.frontend.v3_0.helpers.Eagerly
+import org.neo4j.cypher.internal.frontend.v3_0.notification.InternalNotification
+import org.neo4j.graphdb.{ResourceIterator, _}
+
+import scala.collection.{Map, mutable}
+
+/*
+* An AcceptingExecutionResult delegates all methods to the accept method which is the responsibility
+* of extending classes to provide.
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala]: I suggest `... which extending classes are responsible for providing`.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+  override final def javaColumnAs[T](column: String): ResourceIterator[T] = new WrappingResourceIterator[T] {
+    def hasNext = self.hasNext
+
+    def next() = extractJavaColumn(column, innerIterator.next()).asInstanceOf[T]
+  }
+
+  override def columnAs[T](column: String): Iterator[T] = map { case m => extractColumn(column, m).asInstanceOf[T] }
+
+  override def javaIterator: ResourceIterator[util.Map[String, Any]] = new
+      WrappingResourceIterator[util.Map[String, Any]] {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala]: Break line before `new`.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    val builder = Seq.newBuilder[Map[String, String]]
+    doInAccept(populateDumpToStringResults(builder))
+    formatOutput(writer, columns, builder.result(), queryStatistics())
+  }
+
+  /*
+   * NOTE: This should ony be used for testing, it creates an InternalExecutionResult
+   * where you can call both toList and dumpToString
+   */
+  def toEagerIterableResult(planner: PlannerName, runtime: RuntimeName): InternalExecutionResult = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala:90]: Perhaps this could be moved to some test helper class?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala [hunk 11]
```diff
@@ -271,5 +55,9 @@ class CompiledExecutionResult(taskCloser: TaskCloser,
 
     compiledCode.executionPlanDescription()
   }
-}
 
+  override def queryStatistics() = InternalQueryStatistics()
+
+  //TODO delegate to compiled code once writes are being implemented
+  override def executionType = READ_ONLY
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala:62]: What kind of writes? Writes from procedures?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/ExecutionPlanBuilder.scala [hunk 12]
```diff
@@ -89,7 +89,9 @@ object ExecutablePlanBuilder {
 }
 
 trait ExecutablePlanBuilder {
-  def producePlan(inputQuery: PreparedQuery, planContext: PlanContext, tracer: CompilationPhaseTracer = CompilationPhaseTracer.NO_TRACING,
+
+  def producePlan(inputQuery: PreparedQuery, planContext: PlanContext, tracer:
+  CompilationPhaseTracer = CompilationPhaseTracer.NO_TRACING,
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/ExecutionPlanBuilder.scala]: Bad indentation.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/CallProcedureExecutionPlan.scala [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+      ProcedureExecutionResult(taskCloser, ctx, signature, input, description, planType)
+  }
+
+  private def fail(f: FieldSignature, ctx: QueryContext) = {
+    ctx.close(success = false)
+    throw new InvalidArgumentException(
+      s"""Procedure ${signature.name.name} expected an argument with ${f.name} with type ${f.typ}""")
+  }
+
+  private def description = PlanDescriptionImpl(new Id, "CallProcedure", NoChildren, Seq(), Set.empty)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/CallProcedureExecutionPlan.scala]: Isn't `ProcedureCall` slightly more understandable?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+  * This planner takes on queries that requires no planning such as procedures and schema commands
+  *
+  * @param delegate The plan builder to delegate to
+  */
+case class DelegatingProcedureExecutablePlanBuilder(delegate: ExecutablePlanBuilder) extends ExecutablePlanBuilder {
+
+  override def producePlan(inputQuery: PreparedQuery, planContext: PlanContext, tracer: CompilationPhaseTracer,
+
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala:36]: Remove empty line.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    inputQuery.statement match {
+
+      // CALL foo.bar.baz("arg1", 2)
+      case CallProcedure(namespace, ProcName(name), args) =>
+        val signature = planContext.procedureSignature(ProcedureName(namespace, name))
+        if (args.nonEmpty && args.size != signature.inputSignature.size)
+          throw new InvalidArgumentException(
+            s"""Procedure ${signature.name.name} takes ${signature.inputSignature.size}
+                  |arguments but ${args.size} was provided.""".stripMargin)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala]: Bad indentation.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala (hunk 11 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala:62]: In this case you are commenting on `CompiledExecutionResult` it doesn't support write operations yet, when we do we should fix update this

Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala:62]: Ah, gotcha. I was confused over this part since it moved from last time I read it.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala (hunk 10 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/AcceptingExecutionResult.scala:90]: it needs direct access to `doInAccept`. This is only used to make it possible to create `RewindableExecutionResult` for tests. We should refactor how that works and remove this method instead. (and it wasn't added in this PR just moved around code)

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala (hunk 14 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala:36]: No I like it

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala (hunk 4 above)
Comment (pontusmelke) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:33]: I added these tests. Note though that the built-procedures will not take into account whether the labels, types, or properties are in-use. We will either update these procedures to take _in use_ into account, or we will add that functionality as separate procedures.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala (hunk 5 above)
Comment (pontusmelke) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:250]: Added test for having too many arguments. Tests for adding procedures are done at the kernel level, the only exposed way to create procedures will be from annotated java classes which are loaded when the database boots. The way it is done here is just a hack for tests.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala (hunk 11 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/CompiledExecutionResult.scala:62]: There will soon be coming actual procedures that also does writes, the `executionType` will not be a problem, not exactly sure about `queryStatistics` though bet let's worry about that when we get to updating procedures. So far all procedures are read-only.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala (hunk 3 above)
Comment (pontusmelke) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:31]: done

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala (hunk 14 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/procs/DelegatingProcedureExecutablePlanBuilder.scala:36]: Really? In the middle of the parameter list?

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala (hunk 4 above)
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CallProcedureAcceptanceTest.scala:33]: :+1:
